includes/class-amp-rest-api.php

```diff
@@ -0,0 +1,54 @@
+<?php
+
+/**
+ * AMP REST API.
+ *
+ * @package AMP
+ * @since 2.0
+ */
+
+/**
+ * Class AMP_REST_API.
+ */
+class AMP_REST_API
+{
+
+  /**
+   * Init.
+   */
+  public static function init()
+  {
+    if (!class_exists('AMP_REST_API')) {
+      return;
+    }
+
+    // Register one rest_prepare filter for each custom post type.
+    add_action('registered_post_type', array(__CLASS__, 'add_rest_prepare_filter'));
+  }
+
+  /**
+   * Register a rest_prepare filter for the $post_type.
+   *
+   * @param string $post_type The post type.
+   * @return void
+   */
+  public static function add_rest_prepare_filter($post_type)
+  {
+    add_filter('rest_prepare_' . $post_type, array(__CLASS__, 'add_content_amp_field'), 10, 3);
+  }
+
+  /**
+   * Adds a new field `amp` in the content of the REST API response.
+   *
+   * @param WP_REST_Response $response Response object.
+   * @param WP_Post $post Post object.
+   * @param WP_REST_Request $request Request object.
+   * @return WP_REST_Response Response object.
+   */
+  public static function add_content_amp_field($response, $post, $request)
+  {
+    if (isset($response->data['content']['rendered']))
+      $response->data['content']['amp'] = "An AMP valid HTML generated with the AMP Plugin.";
```

Comment:
To confirm, would this be a full valid AMP document? In other words, would it include the head and body with the body including just the_content()? And the head would include the style [amp-custom]?

Reply:
I'll put together some code to show how to generate this.

However, generating this content is much more expensive than the normal content.rendered field. I think it should potentially be omitted unless explicitly requested.

Reply:
+1 as making the field opt-in instead of being shown by default. For the decoupled consumption use-case I think a full HTML document is not as useful as separate fields for the AMP-valid `body` contents, the tree-shaken CSS, and a list of the necessary scripts. The front-end implementation can then combine those pieces and supplement them with their own, or other data.

Reply:
@luisherranz How are you intending to incorporate this into Frontity? For the Shadow AMP context, having the full AMP document is best.

But if looking to inject the document into another AMP page, then this will require more work. If the document is indeed split into parts for body content, scripts, and styles—then how would they be combined with the elements on the target AMP page?

Reply:
Also, in relation to including the full AMP response among the fields being poor for performance, what if the AMP content were only made available via a link? When doing an `_embed` request then the full AMP content could be included among `_embedded`.

Reply:
Ok, so now this is getting very interesting. We have many options here. I'll try to summarize the options and the decisions we need to make.

### Options for AMP output:
 1. `content.amp`: full document AMP.
 2. Maintain `content.amp` but add new field `amp`:
     - `content.amp`: only `the_content()` converted to AMP
     - `amp.styles`: tree-shaken CSS
     - `amp.scripts`: array of used scripts
3. Put everything in `content.amp`:
    - `content.amp.body`: only `the_content()` converted to AMP
    - `content.amp.styles`: tree-shaken CSS
    - `content.amp.scripts`: array of used scripts
4. Optional using the query (see below).

### Options for the opt-in query:
1. add a separate endpoint and use `_links` to make it embeddable with `_embed`.
1. add a new query, for example `_amp`.
    1. make it boolean (include it or not).
    1. add option for output format, for example:
      - `_amp=full`: full document.
      - `_amp=body`: only `the_content`.

----

The only thing I have strong feelings against is to use `_embed` for the opt-in. If this is expensive, I think it should be a separate param. Most headless themes are probably using `_embed` on each request, so I don't see why they would need `amp` unless they are actively using AMP.

Reply:
Should we just use `register_rest_field` and then opt-in is done via using the `_fields` param?

Reply:
> How are you intending to incorporate this into Frontity?

I think both options suit different use cases:
- Do you have a PWA and want to insert the `content` using AMP instead of HTML because it has some nice performance features? Use the Shadow AMP API.
- Do you want to generate full AMP pages using a decoupled solution like Frontity or Next? Insert the content and add the scripts and styles to your head.

If we were forced to do only one, I'd rather go with the separate body content, scripts, and styles because as @ataylorme said it's easier to combine them to generate the full document than the other way around.

Reply:
I guess one interesting question would be: can the Shadow AMP API be used inside a AMP document?

Reply:
Is this done elsewhere with the `_fields` param? Isn't it normally used only for reducing the number of fields, rather than adding to the number of fields included?

Reply:
Let's move this discussion either to the PR comments or to #1014 because it is getting buried here.

Reply:
Moved to https://github.com/ampproject/amp-wp/issues/1014#issuecomment-512884698